### PR TITLE
Implement regular expression for affiliate links following 'to' or 'at'

### PR DESCRIPTION
**Pull request by margaretchan**
To be matched, links need to begin with either http:// or https:// and contain more than 1 character after. 
ex. "Head to https://google.com for 50% off" --> https://google.com

Punctuation in sentence structure is also considered so the following cases are matched correctly:
ex. "Use my link at https://google.com, click on the search bar, and have fun" --> https://google.com
ex. "Google is a great search engine (go to https://google.com)!" --> https://google.com

### CODE REVIEW

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java [hunk 1]
```diff
@@ -334,5 +345,165 @@ public void codeWithQuotes_multipleMatches() {
         assertThat(actual, equalTo(Arrays.asList("bobat15", "COOLIRPA")));
     }
 
+
+    /*                                                                   *
+     * ============= TEST 'TO' AND 'AT' FOR AFFILATE LINKS ============= *
+     * ex. "go to https://google.com" | "50% off at https://website.com" */
+
+    private final Pattern TO_AT_LINKS_PATTERN = DescriptionParser.Patterns.TO_AT_LINKS.getPattern();
```
Comment (josefj1519) [inline src/test/java/com/google/step/youtube/DescriptionParserTest.java]: Why don't declare it here vs. at the top of the field?

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java [hunk 2]
```diff
@@ -334,5 +345,165 @@ public void codeWithQuotes_multipleMatches() {
...
+        String desc = "Use code LINUS and get 25% off GlassWire at  https://lmg.gg/glasswire";
+
+        List<String> actual = DescriptionParser.findMatches(TO_AT_LINKS_PATTERN, desc);
+        assertThat(actual, equalTo(Collections.emptyList()));
+    }
+
+    @Test
+    public void toAtLinks_newLineBetween() {
+        String desc = "Use code LINUS and get 25% off GlassWire at\nhttps://lmg.gg/glasswire";
+
```
Comment (josefj1519) [inline src/test/java/com/google/step/youtube/DescriptionParserTest.java]: optional: I would remove these blank lines between the description and the actual variable.  I think it hurts readability.

**Review by josefj1519 — APPROVED** [review]

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java (hunk 2 above)
Comment (dallen68) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: +1

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java (hunk 1 above)
Comment (dallen68) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: +1

**Review by dallen68 — APPROVED** [review]

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java (hunk 1 above)
Comment (margaretchan) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: Moved to declare at top of class.

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java (hunk 2 above)
Comment (margaretchan) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: Removed.

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java [hunk 3]
```diff
@@ -329,10 +307,149 @@ public void codeWithQuotes_multipleMatches() {
...
             + "Save 33% on your first Native Deodorant Pack - normally $36, you'll get it for "
             + "$24! Click here: https://bit.ly/nativecoolirpa and use my code \"COOLIRPA\".";
-
         List<String> actual = DescriptionParser.findMatches(CODE_WITH_QUOTES_PATTERN, desc);
         assertThat(actual, equalTo(Arrays.asList("bobat15", "COOLIRPA")));
     }
 
+
+    /*                                                                   *
+     * ============= TEST 'TO' AND 'AT' FOR AFFILATE LINKS ============= *
```
Comment (LiRq95) [inline src/test/java/com/google/step/youtube/DescriptionParserTest.java]: I was thinking, as the other regexes are also run in the actual code, is there a guarantee that only to...at would match and produce the correct string rather than a separate regex giving a false result when validating here? You'll likely want to run minimal cases for single regexes and have more cases where you validate a string for all regexes right?

Comment (margaretchan) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: The tests for single regexs call findMatches() instead of parse() so it only runs only 1 regex. This way I can guarantee that for all tests below line 316, only TO_AT_LINKS is run. The idea was to test specific edge cases for each individual regex and then test them all generally together (these are the tests at the top of the file).

**Review by LiRq95 — APPROVED** [review]

File: src/test/java/com/google/step/youtube/DescriptionParserTest.java (hunk 3 above)
Comment (LiRq95) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: Hmm yeah I'm just wondering if there's some way of organizing the test so it's more readable like input + expected output pairs in lists based on regexes, and then a regex test would just run through those per list, while an overall test would apply all regexes on each entry in all the lists to make sure there's no conflicting overlap...

Kind of like my initial suggestion with having a csv to validate over, but instead of it being in a separate file as discouraged by derrill, just have them be constants defined in the test class. But everything works fine at the moment so it's really just something optional to think about if you want to reuse code more effectively.

Comment (margaretchan) [reply src/test/java/com/google/step/youtube/DescriptionParserTest.java]: Ah I see, but would that method make it less apparent which edge case each input/output pair is testing?
